common/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/common/config/CliConfig.scala

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2018-2021 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.storage.bigquery.common.config
+
+import java.util.Base64
+
+import cats.Monad
+import cats.data.{EitherT, ValidatedNel}
+import cats.implicits.toShow
+import cats.syntax.either._
+import com.monovore.decline.Opts
+import com.typesafe.config.ConfigFactory
+import io.circe.{Decoder, DecodingFailure, Encoder, Json}
+import io.circe.config.parser
+import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
+import io.circe.parser.parse
+
+import com.snowplowanalytics.iglu.client.Resolver
+import com.snowplowanalytics.iglu.client.resolver.{InitListCache, InitSchemaCache}
+import com.snowplowanalytics.snowplow.storage.bigquery.common.config.model._
+
+final case class CliConfig(
+  projectId: String,
+  loader: Config.Loader,
+  mutator: Config.Mutator,
+  repeater: Config.Repeater
+)
+
+object CliConfig {
+  sealed trait Environment {
+    val config: Config
+    val resolverJson: Json
+    val projectId: String
+
+    def getFullSubName(sub: String): String     = s"projects/$projectId/subscriptions/$sub"
+    def getFullTopicName(topic: String): String = s"projects/$projectId/topics/$topic"
+  }
```

Comment:
Here's an interesting thing.  The code base still compiles if you make this a regular trait, not a sealed trait, and if you remove config and resolverJson as trait fields:

```scala
  trait Environment {
    val projectId: String

    def getFullSubName(sub: String): String     = s"projects/$projectId/subscriptions/$sub"
    def getFullTopicName(topic: String): String = s"projects/$projectId/topics/$topic"
  }
```

It might be a sign that sealed trait is not the tool you want for this job.  You are using this trait more as a mixin to add the `getFullSubName` and `getFullTopicName` methods.


You don't have to copy this, but here is a completely alternative approach:

```scala
case class Environment[A](config: A, resolverJson: Json, projectId: String) {
    def getFullSubName(sub: String): String     = s"projects/$projectId/subscriptions/$sub"
    def getFullTopicName(topic: String): String = s"projects/$projectId/topics/$topic"
}

type LoaderEnvironment = Environment[Config.Loader]
type MutatorEnvironment = Environment[Config.Mutator]
type RepeqterEnvironment = Environment[Config.Repeater]
```

Reply:
Hey Ian, that is interesting indeed. I did try at some point making it a parameterised trait but it didn't work out and maybe the reason was that I should have also switched it to be a case class.

Other than succinctness, are there any additional benefits to this alternative approach? I believe my initial attempt at parameterisation was meant to cut down on the number of implicit codecs that needed to be derived, but later the Environment trait got moved out of the model and didn't really need codecs.